packages/iov-core/src/workers/signingservice.worker.ts

```diff
@@ -0,0 +1,198 @@
+/// <reference lib="webworker" />
+
+/// A proof-of-work RPC server for the out pf process signer
+
+import { ChainId, UnsignedTransaction } from "@iov/bcp-types";
+import { bnsConnector } from "@iov/bns";
+import { ethereumConnector } from "@iov/ethereum";
+import {
+  jsonRpcCodeInvalidParams,
+  jsonRpcCodeInvalidRequest,
+  jsonRpcCodeMethodNotFound,
+  jsonRpcCodeServerErrorDefault,
+  JsonRpcErrorResponse,
+  JsonRpcRequest,
+  JsonRpcResponse,
+  parseJsonRpcId,
+  parseJsonRpcRequest,
+} from "@iov/jsonrpc";
+import { Ed25519HdWallet, HdPaths, Secp256k1HdWallet, UserProfile } from "@iov/keycontrol";
+
+import { MultiChainSigner } from "../multichainsigner";
+import { ServerCore } from "../servercore";
+
+interface RpcCallGetIdentities {
+  readonly name: "getIdentities";
+  readonly reason: string;
+  readonly chainIds: ReadonlyArray<ChainId>;
+}
+
+interface RpcCallSignAndPost {
+  readonly name: "signAndPost";
+  readonly reason: string;
+  readonly transaction: UnsignedTransaction;
+}
+
+type RpcCall = RpcCallGetIdentities | RpcCallSignAndPost;
+
+class ParamsError extends Error {}
+class MethodNotFoundError extends Error {}
+
+function isArrayOfStrings(array: ReadonlyArray<any>): array is ReadonlyArray<string> {
+  return array.every(element => typeof element === "string");
+}
+
+function parseRpcCall(data: JsonRpcRequest): RpcCall {
+  const params: ReadonlyArray<unknown> = data.params;
+
+  switch (data.method) {
+    case "getIdentities": {
+      const reason = params[0];
+      const chainIds = params[1];
+      if (typeof reason !== "string") {
+        throw new ParamsError("1st parameter (reason) must be a string");
+      }
+      if (!Array.isArray(chainIds)) {
+        throw new ParamsError("2nd parameter (chainIds) must be an array");
+      }
+      if (!isArrayOfStrings(chainIds)) {
+        throw new ParamsError("Found non-string element in chainIds array");
+      }
+      return {
+        name: "getIdentities",
+        reason: reason,
+        chainIds: chainIds,
+      };
+    }
+    case "signAndPost": {
+      const reason = params[0];
+      const transaction = params[1];
+      if (typeof reason !== "string") {
+        throw new ParamsError("1st parameter (reason) must be a string");
+      }
+      if (typeof transaction !== "object") {
+        throw new ParamsError("2nd parameter (transaction) must be an object");
+      }
+      return {
+        name: "signAndPost",
+        reason: reason,
+        transaction: transaction as UnsignedTransaction,
+      };
+    }
+    default:
+      throw new MethodNotFoundError("Unknown method name");
+  }
+}
+
+const bnsdUrl = "ws://localhost:22345";
+const bnsdFaucetMnemonic = "degree tackle suggest window test behind mesh extra cover prepare oak script";
+const ethereumUrl = "http://localhost:8545";
+const ganacheMnemonic = "oxygen fall sure lava energy veteran enroll frown question detail include maximum";
+
+async function main(): Promise<void> {
```

Comment:
This is a bit odd to me... Above looked like production code. This looks like test code. Why in same file (I assume something about webworker build process)

Reply:
The (Web)Worker is it's own JS entry file and this file only exists for tests/demo.

However, we sould extract much of this test file into a reusable JSON-RPC wrapper around ServerCore (https://github.com/iov-one/iov-core/issues/650)